build-check-comments.sh

```diff
@@ -0,0 +1,105 @@
+#!/usr/bin/env bash
+
+set -eu
+
+# Reliable cross-platform comment checker for go files.
+# Uses docker and stable linux distro for reliable grep regex and
+# to make sure we work on Linux and OS X and oh even Windows! :)
+# Main issue is minor incompatibies between grep on various platforms.
+#
+# Linux: set USE_DOCKER=yes to force run in docker
+# All other platforms: uses docker by default. Set USE_DOCKER=no to override.
+#
+# Checks Go code comments if they have trailing period (excludes protobuffers and vendor files).
+# Comments with more than 3 spaces at beginning are omitted from the check, example: '//    - foo'.
+# This does not include top-level commments for funcs and types.
+#
+# Example:
+#   func main() {
+#	    // comment without period, will trigger check
+#	    //comment without leading space, will trigger check.
+#	    // comment without trailing space, will trigger check.
+#	    // good comment, leading space, ends with period, no trailing space.
+#     //     - more than 3 leading spaces, will pass
+#	    app := kingpin.New(filepath.Base(os.Args[0]), "A block storage based long-term storage for Prometheus")
+#   }
+
+
+# Dir of this script
+declare ROOT_DIR=$(cd $(dirname ${BASH_SOURCE}) && pwd)
+declare THIS_SCRIPT="${ROOT_DIR}/$(basename "${BASH_SOURCE}")"
+
+# Image to use if we do docker-based commands, centos is most stable right?
+declare IMAGE="centos:7"
+
+# User can explicitly ask to run in docker
+declare USE_DOCKER=${USE_DOCKER:=""}
+
+# For OS X, always use Docker as we have nasty
+# compat GNU/BSG issues with grep.
+if test "Linux" != "$(uname || true)"
+then
+  # Allow overriding for non-linux platforms
+  if test "no" != "${USE_DOCKER}"
+  then
+    USE_DOCKER="yes"
+  fi
+fi
+
+if test "yes" == "${USE_DOCKER}"
+then
+    # Make sure we only attach TTY if we have it, CI builds won't have it.
+    declare TTY_FLAG=""
+    if [ -t 1 ]
+    then
+        TTY_FLAG="-t"
+    fi
+
+    # Annoying issue with ownership of files in mapped volumes.
+    # Need to run with same UID and GID in container as we do
+    # on the machine, otherwise all output will be owned by root.
+    # Doesn't happen on OS X but does on Linux. So we will do
+    # UID and GID for Linux only (this won't work on OS X anyway).
+    declare USER_FLAG=""
+    if test "Linux" == "$(uname || true)"
+    then
+        USER_FLAG="-u $(id -u):$(id -g)"
+    fi
+
+    (
+      set -x
+      docker run \
```

Comment:
Maybe we could add a check if docker is available and if not, bail out with comment why it is needed?

Reply:
currently we get this if docker isn't there:

> ./build-check-comments.sh: line 71: docker: command not found

is this good enough as it is? 

otherwise we'd rely in `which` which incidentally also may not be present, like in `centos:7` docker image for example and which may lead to false positives (have `docker` no `which`). :) Case in point, running in `centos:7` base image:

```
[root@98cad6a96d96 thanos]# yum install -y make git
[root@98cad6a96d96 thanos]# make build
make: which: Command not found
No git binary found.
make: *** [check-git] Error 1
```

I will add a `printf` saying `Make sure Docker is installed if you you get errors`

Reply:
Not using which, but added `printf`, is this better?

```
Mayas-MacBook-Pro:thanos philip$ make check-comments
>> checking Go comments trailing periods

 This will run in Docker.
 If you get error, ensure Docker is installed.

+ docker run ...
```

Reply:
Ok, true about the checking that it would bring another complexity. Ok from me :+1: